packages/core/saved-objects/core-saved-objects-migration-server-internal/src/actions/update_source_mappings_properties.ts

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { omit } from 'lodash';
+import * as TaskEither from 'fp-ts/lib/TaskEither';
+import { pipe } from 'fp-ts/lib/pipeable';
+import type { ElasticsearchClient } from '@kbn/core-elasticsearch-server';
+import type { IndexMapping } from '@kbn/core-saved-objects-base-server-internal';
+import { diffMappings } from '../core/build_active_mappings';
+import { type Aliases, versionMigrationCompleted } from '../model/helpers';
+import type { RetryableEsClientError } from './catch_retryable_es_client_errors';
+import { updateMappings } from './update_mappings';
+import type { IncompatibleMappingException } from './update_mappings';
+
+/** @internal */
+export interface UpdateSourceMappingsPropertiesParams {
+  client: ElasticsearchClient;
+  aliases: Aliases;
+  currentAlias: string;
+  versionAlias: string;
+  sourceIndex: string;
+  sourceMappings: IndexMapping;
+  targetMappings: IndexMapping;
+}
+
+/** @internal */
+export enum UpdateSourceMappingsPropertiesResult {
+  Compatible = 'compatible',
+  Updated = 'updated',
+  Incompatible = 'incompatible',
+}
+
+/**
+ * This action performs the following operations:
+ *   - Checks the source mappings for any changes.
+ *   - If there are no changes and the version migration is completed, it returns `Updated`.
+ *   - If there are no changes and the version migration is incomplete, it returns `Compatible`.
+ *   - If there are changes it tries to patch the source mappings.
+ *   - If the patch is successful and the version migration is completed, it returns `Updated`.
+ *   - If the patch is successful and the version migration is incomplete, it returns `Compatible`.
+ *   - If the patch is unsuccessful and the version migration is incomplete, it returns `Incompatible`.
+ *   - If the patch is unsuccessful and the version migration is completed, it returns an error. It is likely an incompatible change caused by the newly enabled plugin or code change.
```

Comment:
This feels like mostly pure synchronous business logic that we want to separate from the async side effects in the actions.

Reply:
I've refactored that part to move this logic to the finite state machine.